### PR TITLE
Global sandboxing setting

### PR DESCRIPTION
**Pull request by aalekseyev**
- Add a new config option --sandbox that lets user control whether or not dune sandboxes actions by default.
- The actions no longer give a boolean "sandbox" or "not sandbox", they can now additionally say "no_special_requirements". The framework allows specifying things like "this command can't work with symlinks, but it can work without sandbox or with copy-based sandbox, but we don't use that right now.
- Annotated some actions in dune with `no_sandboxing` so that all tests keep passing with `--sandbox=copy`
- Tested on ocaml-platform and made sure the compilation succeeds with `--sandbox=copy` too

### CODE REVIEW

**Comment by aalekseyev** [comment]
After talking to @diml, we agreed that we should also add to the dune language the ability to opt out or opt in to sandboxing per rule before advertising this feature to the public.

I intend to do that soon.

File: src/build_system.ml [hunk 1]
```diff
@@ -1366,6 +1371,34 @@ end = struct
...
+        | None, { none = true; _ } ->
+          Some None
+        | Some Sandbox_mode.Copy, { copy = true; _ } ->
+          Some (Some Sandbox_mode.Copy)
+        | Some Symlink, { symlink = true; copy; _ } ->
+          (if copy then
+             Some (Some (if Sys.win32 then Copy else Symlink))
+           else
+             Code_error.raise
+               "This rule requires sandboxing with symlinks, but that won't \
```
Comment (diml) [inline src/build_system.ml]: It seems like we should use `User_error.raise ~loc` here. Otherwise it's not clear which rule we are talking about

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/module_compilation.ml [hunk 2]
```diff
@@ -46,7 +46,10 @@ let build_cm cctx ~dep_graphs ~precompiled_cmi ~cm_kind (m : Module.t) =
         (* symlink the .cmi into the public interface directory *)
         if Module.visibility m <> Visibility.Private
         && (Obj_dir.need_dedicated_public_dir obj_dir) then
-          SC.add_rule sctx ~sandbox:false ~dir
+          SC.add_rule sctx
+            (* CR-someday aalekseyev: why do we have [no_sandboxing] here? *)
```
Comment (diml) [inline src/module_compilation.ml]: We are just doing a syscall without calling out to an external command. We can reasonably expect that sandboxing is not necessary. Though, it seems fine to remove the argument altogether

File: src/sandbox_config.ml [hunk 3]
```diff
@@ -0,0 +1,129 @@
+open! Stdune
+
+(* ['a t] represents a total map from [Sandbox_mode.t] to ['a] *)
```
Comment (diml) [inline src/sandbox_config.ml]: What about calling this `Sandbox_mode.Dict.t`? This is what we do in other modules such as `Cm_kind`, `Mode`, ...

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/module_compilation.ml (hunk 2 above)
Comment (aalekseyev) [reply src/module_compilation.ml]: I added a function `Action.is_useful_to_sandbox` to help with that.

File: src/sandbox_config.ml (hunk 3 above)
Comment (aalekseyev) [reply src/sandbox_config.ml]: Thanks, that's obviously better!

**Comment by aalekseyev** [comment]
I found an annoying bug: promotion doesn't work when you sandbox.
The reason is that when `Action_exec` tries to interpret the `Diff` stanza, it finds the sandboxed versions of all paths, and therefore it gets confused. For a start, `is_copied_from_source_tree` doesn't work, but I suspect that's not the only issue.

Relatedly, I suspect "optional" dependencies won't work when the sandbox mode is "copy". (the "Copy" probably requires the src to exist)

**Comment by diml** [comment]
What about "unsandboxing" the paths when the files don't match?

For optional dependencies, are you thinking of `diff?`? This was intended to be used as follow:

```
(progn
 (run cmd ...)
 (diff? ...))
```

the idea being that `cmd` might optionally generate a correction. That scheme should work with sandboxing, although we probably need to remember to copy back the optional files from the sandbox.

**Comment by aalekseyev** [comment]
Ah, I see. I misunderstood what "optional" meant here.

File: src/build_system.ml [hunk 4]
```diff
@@ -825,6 +826,7 @@ end = struct
                       in
                       let rule =
                         Pre_rule.make ~locks ~context:(Some context) ~env
+                          (* [no_sandboxing] here is necessary for some reason *)
```
Comment (diml) [inline src/build_system.ml]: This comment looks stale

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by aalekseyev** [comment]
- I fixed promotion partially (just the path handling, not the optional target handling).
- I added something to the doc.
- I tested odoc build and indeed it seems the alias dependency failure is fixed.
- I fixed the line directive stuff.

I don't want to test the optional targets thing in this PR because it's complicated and the PR is already big.

**Review by ghost — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**